1단계 - 엔티티 매핑

popotec opened the pull request:
안녕하세요!
JPA 1단계 미션 코드리뷰 요청드립니다.
잘 부탁드립니다🙂

popotec committed:
docs(README.md): 미션 정의

popotec committed:
chore: 설정파일 변경

- sql dialect를 Mysql로 변경
- JPA 실행시 sql 출력

popotec committed:
feat(BaseTimeEntity): 공통적인 생성시간, 수정시간 필드를 분리

popotec committed:
feat(Answer): Answer Entity 정의

popotec committed:
feat(Answer): equals, hashcode를 재정의하여 Answer Entity 비교

popotec committed:
feat(AnswerRepository): 학습테스트 구현

popotec committed:
chore: 불필요 annotation 삭제

popotec committed:
chore: 설정파일 db 테이블 생성 추가

popotec committed:
feat(Question): Question entity 생성

popotec committed:
feat(DeleteHistory): DeleteHistory entity 생성

popotec committed:
feat(User): User entity 생성

popotec committed:
test(User): 사용자 계정 정보 비교 테스트

- matchPassword
- equalsNameAndEmail

popotec committed:
chore: 클래스 멤버 변수와 인스턴스 멤버 변수 선언 순서 변경

popotec committed:
test(DeleteHistoryTest): DeleteHistory Entity 테스트

popotec committed:
chore(DeleteHistory): 불필요한 초기화 제거

popotec committed:
feat(DeleteHistoryRepository): 학습테스트

popotec committed:
chore: 테스트 패키지 이동

popotec committed:
chore: repository로 패키지 이동

popotec committed:
chore: spring data jpa에서 자동 생성해주는 메서드 삭제

- findById()
- findByAll()

popotec committed:
test(QuestionRepository): QuestionRepository 학습테스트

popotec committed:
chore: 미사용 import 삭제

popotec committed:
chore: 패키지 구조 변경에 따른 import 추가

popotec committed:
test(UserRepository): UserRepository 학습테스트

na-mul commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -5,89 +5,117 @@
+	@Id
+	@GeneratedValue
+	private Long id;
+	private String contents;
```
`contents`를 그냥 String으로 만들면 요구사항이었던 clob을 충족시키지 못하게 됩니다.
- 수정을 부탁드릴께요

na-mul commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -5,89 +5,117 @@
+	@GeneratedValue
+	private Long id;
+	private String contents;
+	private boolean deleted;
```
- not null 조건을 고려하여 작성 부탁드릴께요!

na-mul commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -5,89 +5,117 @@
+
+	public void setId(Long id) {
+		this.id = id;
+	}
```
- 기본적으로 존재했던 코드들, 특히 `setter` 메소드들은 정리를 부탁드릴께요!

na-mul commented on src/main/java/qna/domain/BaseTimeEntity.java:
```diff
@@ -0,0 +1,31 @@
+import org.springframework.data.annotation.LastModifiedDate;
+import org.springframework.data.jpa.domain.support.AuditingEntityListener;
+
+@EntityListeners(AuditingEntityListener.class)
+@MappedSuperclass
```
Auditing 기능을 잘 활용해주셨습니다!👍
다만 이 기능이 실제로 동작하기 위해서는 
`public class Application`에 `@EnableJpaAuditing` 어노테이션을 포함시켜주셔야 할 것 같아요.
- 수정을 부탁드릴께요!

na-mul commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -3,12 +3,22 @@
     private Long id;
-    private ContentType contentType;
     private Long contentId;
+    private ContentType contentType;
```
`enum` 타입의 경우 역시 어노테이션을 붙여주어야 합니다.
- `@Enumerated(EnumType.STRING)` 어노테이션 사용을 검토 부탁드릴께요

na-mul commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -17,15 +27,35 @@ public DeleteHistory(ContentType contentType, Long contentId, Long deletedById,
+            contentType == that.contentType &&
+            Objects.equals(contentId, that.contentId) &&
+            Objects.equals(deletedById, that.deletedById);
     }
```
- IDE가 기본으로 생성해준 `equals` 메소드라 할지라도 코드 컨벤션을 지켜주시는 방향으로 수정 부탁드릴께요!
- 그리고 비슷한 상황으로 보이는 `equals` 메소드들도 한번에 수정 부탁드릴께요

na-mul commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,83 +1,114 @@
+	private Long id;
+	private String contents;
+	private String title;
+	private boolean deleted;
```
- 모두 적절한 어노테이션을 붙여주시면 감사하겠습니다! 🙏

na-mul commented on src/main/java/qna/repository/QuestionRepository.java:
```diff
@@ -1,10 +1,12 @@
-package qna.domain;
+package qna.repository;
```
패키지 정리 감사합니다.
팀의 상황에 따라서 `repository`들이 `domain` 패키지 하위로 가기도 하니 그 점은 참고 부탁드릴께요!

na-mul commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,48 @@
+    @Autowired
+    EntityManager em;
+
+    @Test
+    public void 엔티티_생성(){
+        Answer answer = new Answer(UserTest.JAVAJIGI, QuestionTest.Q1, "Answer entity unit test");
+        em.persist(answer);
+        Answer findAnswer = em.createQuery("select a from Answer a where a.id=:id",Answer.class)
+            .setParameter("id",answer.getId())
+            .getSingleResult();
+        assertThat(findAnswer).isEqualTo(answer);
+    }
```
JPQL을 잘 학습해주셨네요! 👍😄

na-mul commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,85 @@
+		assertThat(findAnswer.isPresent()).isTrue();
+	}
+
+	@Test
+	void findByWriterIdAndQuestionId() {
+		Answer savedAnswer = answerRepository.save(
+			new Answer(UserTest.JAVAJIGI, QuestionTest.Q1, "Answer entity unit test"));
+		List<Answer> findAnswers = answerRepository.findByWriterIdAndQuestionId(UserTest.JAVAJIGI.getId(),
+			QuestionTest.Q1.getId());
+		assertThat(findAnswers.contains(savedAnswer));
+	}
+
+	@Test
+	void findByContentsStartingWith() {
+		String content="Answer entity unit test";
+		Answer savedAnswer = answerRepository.save(
+			new Answer(UserTest.JAVAJIGI, QuestionTest.Q1, content));
+		List<Answer> findAnswers = answerRepository.findByContentsStartingWith(content);
+		assertThat(findAnswers.contains(savedAnswer));
+	}
+
+	@Test
+	void findByCreatedAtBetween() {
+		String content="Answer entity unit test";
+		Answer savedAnswer = answerRepository.save(
+			new Answer(UserTest.JAVAJIGI, QuestionTest.Q1, content));
+		List<Answer> findAnswers = answerRepository.findByCreatedAtBetween(LocalDateTime.now().minusMinutes(30),LocalDateTime.now().plusMinutes(30));
+		assertThat(findAnswers.contains(savedAnswer));
+	}
```
`assertThat` 구문이 올바르게 종료되지 않은 것 같아요! 
또한 `findByCreatedAtBetween` 테스트의 경우, 위 말씀드린 사유로 createdAt이 정상적으로 세팅되어있지 않아요.
- 전반적인 수정을 부탁드립니다

popotec replied:
안녕하세요 동환님🙂
엔티티 요구사항을 꼼꼼히 반영하지 못했던 것 같습니다.
코멘트 주신 것 참고하여 부족한 부분들을 보완해보았습니다.
감사합니다😀

na-mul reviewed (changes requested):
안녕하세요! JPA 미션의 리뷰를 맡게된 노동환입니다!
올려주신 내용 잘 봤습니다! 고생 많으셨습니다. 🙇‍♂️
조금 더 살펴봤으면 하는 부분들을 코멘트로 남겨드렸어요.
확인해주셔서 다시 request 부탁드릴게요!

popotec committed:
fix(Auditing): @EntityListeners(AuditingEntityListener.class) 실행을 위한 어노테이션 추가

popotec committed:
fix(User): Entity 수정

- 제약조건 추가
  - length 제약조건 : email, name, password, userId
  - not null 제약조건 : name, password, userId
- set 메서드 삭제
- Equals, toString 수정

popotec committed:
fix(Question): Entity 수정

- CLOB 타입 저장을 위한 어노테이션(@Lob) 추가
- 제약조건 추가
  - length 제약조건 : title
  - not null 제약조건 : title, deleted
- set 메서드 삭제
  - delete는 메세지를 보내 삭제하도록 변경
- Equals, toString 수정

popotec committed:
fix(Answer): Entity 수정

- CLOB 타입 저장을 위한 어노테이션(@Lob) 추가
- 제약조건 추가
  - not null 제약조건 : deleted
- set 메서드 삭제
  - delete는 메세지를 보내 삭제하도록 변경
- Equals, toString 수정

popotec committed:
fix(BaseTimeEntity): not null 제약조건 추가

popotec committed:
fix(DeleteHistory): Entity 수정

- enum 타입 어노테이션 추가 : @Enumerated(EnumType.STRING)
- Equals 수정

na-mul commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -5,89 +5,117 @@
+		deleted = true;
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+
+		Answer answer = (Answer)o;
+		return isDeleted() == answer.isDeleted() && Objects.equals(getId(), answer.getId())
+			&& Objects.equals(getContents(), answer.getContents()) && Objects.equals(getQuestionId(),
+			answer.getQuestionId()) && Objects.equals(getWriterId(), answer.getWriterId())
+			&& Objects.equals(getCreatedAt(), answer.getCreatedAt()) && Objects.equals(getUpdatedAt(),
+			answer.getUpdatedAt());
+	}
```
앞서 코멘트를 하나 놓친 부분이 있네요!

entity의 equals 메소드를 오버라이딩할 때는 주로 비즈니스 키를 활용합니다. 
보통 중복되지 않고 거의 변하지 않는 필드들이 활용됩니다. DB로 치면 후보키들을 사용한다고 볼 수 있을 것 같아요.
따라서 contents나 updatedAt 등은 적절치 않습니다. 

그러나 만약에 값 타입 ( `@Embedded` ) 이라면 모든 필드를 equals 메소드로 비교하곤 하니 참고 부탁드려요!

na-mul reviewed (approved):
안녕하세요 지희님, 요구사항에 맞게 잘 수정해주셨네요!
approve 하도록 하겠습니다!
참고하면 좋을 간단한 코멘트도 하나 달아두었으니 괜찮으시면 다음 단계 진행 하실 때 고려해주세요! 🙇‍♂️
다음 단계도 화이팅입니다. 😄
